File: week1/BOJ_10026_적록색약/seokjun_10026.js
class Queue {
  items = [];
  front = 0;
  rear = 0;

  size() {
    return this.rear - this.front;
  }

  push(item) {
    this.items.push(item);
    this.rear += 1;
  }

  pop() {
    if (this.size() <= 0) return null;

    return this.items[this.front++];
  }
}

const readline = require("readline");
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

const input = [];
let currentLine = 0;

rl.on("line", (line) => {
  input.push(line);
});

rl.on("close", () => {
  const N = Number(input[currentLine++]);
  const picture = Array.from({ length: N }, () => {});

  for (let n = 0; n < N; n++) {
    picture[n] = input[currentLine++].split("");
  }

  let normal = 0;
  let abnormal = 0;

  let visitedNormal = Array.from({ length: N }, () => {
    return Array.from({ length: N }, () => false);
  });

  let visitedAbnormal = Array.from({ length: N }, () => {
    return Array.from({ length: N }, () => false);
  });

  const dirs = [1, -1, 0, 0, 1, -1];
  const bfsNormal = (i, j) => {
    const q = new Queue();
    q.push([i, j]);
    visitedNormal[i][j] = true;

    while (q.size() > 0) {
      const [y, x] = q.pop();
      const currentColor = picture[y][x];
      for (let k = 0; k < 4; k++) {
        const [newY, newX] = [y + dirs[k], x + dirs[k + 2]];

        if (
          newY < 0 ||
          newX < 0 ||
          newY >= N ||
          newX >= N ||
          visitedNormal[newY][newX] ||
          currentColor !== picture[newY][newX]
        ) {
          continue;
        }

        q.push([newY, newX]);
        visitedNormal[newY][newX] = true;
      }
    }
  };

  const bfsAbnormal = (i, j) => {
    const isSameColor = (a, b) => {
      if (a === b || (a === "R" && b === "G") || (a === "G" && b === "R")) {
        return true;
      } else {
        return false;
      }
    };
    const q = new Queue();
    q.push([i, j]);
    visitedAbnormal[i][j] = true;

    while (q.size() > 0) {
      const [y, x] = q.pop();
      const currentColor = picture[y][x];
      for (let k = 0; k < 4; k++) {
        const [newY, newX] = [y + dirs[k], x + dirs[k + 2]];

        if (
          newY < 0 ||
          newX < 0 ||
          newY >= N ||
          newX >= N ||
          visitedAbnormal[newY][newX] ||
          !isSameColor(currentColor, picture[newY][newX])
        ) {
          continue;
        }

        q.push([newY, newX]);
        visitedAbnormal[newY][newX] = true;
      }
    }
  };

  for (let i = 0; i < N; i++) {
    for (let j = 0; j < N; j++) {
      if (!visitedNormal[i][j]) {
        bfsNormal(i, j);
        normal += 1;
      }
    }
  }

  for (let i = 0; i < N; i++) {
    for (let j = 0; j < N; j++) {
      if (!visitedAbnormal[i][j]) {
        bfsAbnormal(i, j);
        abnormal += 1;
      }
    }
  }

  console.log(`${normal} ${abnormal}`);
  process.exit();
});

File: BOJ/DFSBFS/10026_적록색약/seokjun_10026.py
from collections import deque

N = int(input())
picture = []

for i in range(N):
  row = list(input())
  picture.append(row)

normal = 0
abnormal = 0
visitedNormal = [[False for _ in range(N)] for _ in range(N)]
visitedAbnormal = [[False for _ in range(N)] for _ in range(N)]

dirs = [1, -1, 0, 0, 1, -1]

def bfsNormal(i ,j):
  q = deque()
  q.append((i, j))
  visitedNormal[i][j] = True

  while(q):
    cell = q.popleft()
    y = cell[0]
    x = cell[1]
    currentColor = picture[y][x]

    for k in range(4):
      newY = y + dirs[k]
      newX = x + dirs[k+2]

      if (
        newY < 0 or newX < 0 or 
        newY >= N or newX >= N or 
        visitedNormal[newY][newX] or
        currentColor != picture[newY][newX]):
        continue

      q.append((newY, newX))
      visitedNormal[newY][newX] = True;

def bfsAbnormal(i ,j):
  def isSameColor(a, b):
    if (a == b or (a == "R" and b == "G") or (a == "G" and b == "R")):
      return True
    else:
      return False
  
  q = deque()
  q.append((i, j))
  visitedAbnormal[i][j] = True

  while(q):
    cell = q.popleft()
    y = cell[0]
    x = cell[1]
    currentColor = picture[y][x]

    for k in range(4):
      newY = y + dirs[k]
      newX = x + dirs[k+2]

      if (
        newY < 0 or newX < 0 or 
        newY >= N or newX >= N or 
        visitedAbnormal[newY][newX] or
        not isSameColor(currentColor, picture[newY][newX])):
        continue

      q.append((newY, newX))
      visitedAbnormal[newY][newX] = True

for i in range(N):
  for j in range(N):
    if not visitedNormal[i][j]:
      bfsNormal(i, j)
      normal += 1
  
for i in range(N):
  for j in range(N):
    if not visitedAbnormal[i][j]:
      bfsAbnormal(i, j)
      abnormal += 1

print(f'{normal} {abnormal}')
